### DISCUSSION TITLE
Opcodes that always go together

### DISCUSSION THREAD

**Discussion by sweeneyde** [faster-cpython/ideas#323] [Ideas]
Using stats like https://github.com/python/cpython/pull/31957 , I found that on a complete run of `python -m test`, we have:

- ASYNC_GEN_WRAP always goes to YIELD_VALUE
- BEFORE_ASYNC_WITH always goes to GET_AWAITABLE
- BINARY_OP_INPLACE_ADD_UNICODE always goes to STORE_FAST
- CALL always comes from PRECALL
- DICT_MERGE *almost* always (but not always) goes to CALL_FUNCTION_EX
- GET_AWAITABLE *almost* always (but not always) goes to LOAD_CONST
- GET_LEN always goes to LOAD_CONST
- GET_YIELD_FROM_ITER always goes to LOAD_CONST
- IMPORT_NAME always comes from LOAD_CONST
- IMPORT_STAR always comes from IMPORT_NAME
- JUMP_NO_INTERRUPT always goes to SEND
- LIST_TO_TUPLE *almost* always (but not always) goes to LIST_EXTEND
- LOAD_BUILD_CLASS always comes from PUSH_NULL
- MATCH_CLASS always comes from LOAD_CONST
- MATCH_MAPPING always goes to POP_JUMP_IF_FALSE
- MATCH_SEQUENCE always goes to POP_JUMP_IF_FALSE
- PRECALL always goes to CALL
- PRECALL_NO_KW_LIST_APPEND *almost* always (but not always) goes to POP_TOP
- PREP_RERAISE_STAR always comes from LIST_APPEND
- PREP_REREAISE_STAR always goes to COPY
- UNPACK_SEQUENCE_LIST *almost* always (but not always) goes to STORE_FAST_STORE_FAST
- WITH_EXCEPT_START always comes from PUSH_EXC_INFO

Maybe some of these can use JUMP_TO_INSTRUCTION to eliminate a dispatch, and maybe some instructions can be combined, whether that's at compile time or specialization time.

**Comment by markshannon** [comment]
There are fives cases for pair A B:
* A always precedes B:  Probably nothing we can do with this.
* A mostly precedes B: Ditto.
* B always follows A: Use JUMP_TO_INSTRUCTION to skip dispatch.
* B usually follows A: Use PREDICT macros to reduce DISPATCH overhead (on Windows).
* A always precedes B *and* B always follows A: Merge them into a single instruction. Does this happen at all except for calls?

And one more case, that we already handle:
* A B is common: Make it a superinstruction.


`CALL` and `PRECALL` exist because we want two step specialization for some calls. So that's a special case.

**Comment by markshannon** [comment]
One other thing. If instructions are rare, like `PREP_REREAISE_STAR`, then we don't want to make them larger or replicate them.
